### PR TITLE
10984:Improve TagsField of keywords

### LINKED ISSUE #10984: Improve TagsField of keywords
This is a follow-up of https://github.com/JabRef/jabref/pull/10910

For things are left open:

1. On paste: If it is possible to add this keyword, directly add it without requiring the user to press Enter
2. Double click on keyword: remove it and enter edit mode. Example:
   double click on "Where" 
image
    leads to 
    image
3. Reorder keywords by drag and drop
4. Alphabetic ordering of keywords. In case the list is already alphabetically ordered, add the new keyword in alphabetical order. Otherwise, add it to the end

If something does not work, it should be explained why it is not possible.

### PR DESCRIPTION
**Pull request by MrunalKashid02**
Fixes #10984

### Mandatory checks

- [ ] Change in `CHANGELOG.md` described in a way that is understandable for the average user (if applicable)
- [ ] Tests created for changes (if applicable)
- [x] Manually tested changed features in running JabRef (always required)
- [ ] Screenshots added in PR description (for UI changes)
- [ ] [Checked developer's documentation](https://devdocs.jabref.org/): Is the information available and up to date? If not, I outlined it in this pull request.
- [ ] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not, I created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, I submitted a pull request to the documentation repository.

### CODE REVIEW

**Comment by Siedlerchr** [comment]
You need to merge upstream/main in your branch and resolve the merge conflicts

**Comment by MrunalKashid02** [comment]
Merge conflicts because draft PR? or else

**Comment by Siedlerchr** [comment]
Merge conflicts happen when the same code was edited by different people: 
https://www.atlassian.com/git/tutorials/using-branches/merge-conflicts

You need to merge the upstream/main into your branch and then you can use Intellij to resolve them:
https://www.jetbrains.com/help/idea/resolve-conflicts.html

**Review by subhramit — COMMENTED** [review]
Hi Mrunal. May I ask which AI you used to generate the code?

**Comment by MrunalKashid02** [comment]
Is there anything wrong?

**Comment by subhramit** [comment]
> Is there anything wrong?

Not with the fact that you considered contributing, but we don't appreciate it when PRs submitted have majority of the code generated by AI. It increases our effort to review and the candidate's effort to gather context about their own code and find fixes. 

At JabRef, we strive to provide valuable learning experiences to contributors to help them grow their skills. The team invests significant time and resources, with a 2-to-1 ratio of senior developers mentoring each student. When a contribution lacks genuine effort, it not only affects maintainers' time but also takes away opportunities from other students who are eager to learn and contribute meaningfully.

**Comment by MrunalKashid02** [comment]
I apologise for my mistake. will be ok if I again  start by but it requires some time for each task because this is my first time for contribution. Once again sorry.

**Comment by MrunalKashid02** [comment]
Please can if you give me some time to work on this issue.

**Comment by subhramit** [comment]
> I apologise for my mistake. will be ok if I again  start by but it requires some time for each task because this is my first time for contribution. Once again sorry.

It's alright. Try to understand what the code in your PR does properly and see if you can continue. If you are not confident about finishing it, you can ask to be unassigned and pick smaller issues to start with.

**Comment by MrunalKashid02** [comment]
There are 1-4 task. I can do each task one by one this are small task I guess. if you feel that this issue is resolved quickly you may unassigned me. I really appreciate guidance and support.

**Comment by subhramit** [comment]
> There are 1-4 task. I can do each task one by one this are small task I guess. if you feel that this issue is resolved quickly you may unassigned me. I really appreciate guidance and support.

If you can complete any sub-task of the issue fully, that's also fine.

File: src/main/java/org/jabref/gui/fieldeditors/KeywordsEditor.java [hunk 1]
```diff
@@ -86,16 +94,89 @@ public KeywordsEditor(Field field,
...
+        Clipboard clipboard = Clipboard.getSystemClipboard();
+        if (clipboard.hasString()) {
+            String pastedText = clipboard.getString().trim();
+
+            if (!pastedText.isEmpty()) {
+                // Split keywords by the separator and add them
+                String[] keywords = pastedText.split(String.valueOf(viewModel.getKeywordSeparator()));
+
+                for (String keywordText : keywords) {
+                    Keyword keyword = viewModel.getStringConverter().fromString(keywordText.trim());
```
Comment (Siedlerchr) [inline src/main/java/org/jabref/gui/fieldeditors/KeywordsEditor.java:122]: Use Keyword.of

File: src/main/java/org/jabref/gui/fieldeditors/KeywordsEditor.java [hunk 2]
```diff
@@ -86,16 +94,89 @@ public KeywordsEditor(Field field,
...
+
+                    // Add keyword if it's valid and not already present
+                    if (keyword != null && !keywordTagsField.getTags().contains(keyword)) {
+                        keywordTagsField.getTags().add(keyword);
```
Comment (Siedlerchr) [inline src/main/java/org/jabref/gui/fieldeditors/KeywordsEditor.java:126]: You should not operate here on the keywordTagsField directly, instead use the viewModel.keywordListProperty
and simply call add (it already performs a contains check)

**Review by Siedlerchr — CHANGES_REQUESTED** [review]

File: src/main/java/org/jabref/gui/fieldeditors/KeywordsEditor.java (hunk 2 above)
Comment (MrunalKashid02) [reply src/main/java/org/jabref/gui/fieldeditors/KeywordsEditor.java:126]: ok i will make changes

**Comment by MrunalKashid02** [comment]
I am remain with some changes . I will commit changes tomorrow. Will it fine ?

**Comment by koppor** [comment]
> I am remain with some changes . I will commit changes tomorrow. Will it fine ?

Sure. We just need to know whether there is any more activity intended. Otherwise, we will close the PR and look for another contributor.

**Comment by MrunalKashid02** [comment]
I am getting below error
Screenshot 2024-10-16 144452

**Review by koppor — CHANGES_REQUESTED** [review]

File: src/main/java/org/jabref/gui/fieldeditors/KeywordsEditor.java [hunk 3]
```diff
@@ -86,16 +93,84 @@ public KeywordsEditor(Field field,
                 event.consume();
             }
         });
+        keywordTagsField.getEditor().setOnKeyPressed(event -> {
+            if (event.isControlDown() && event.getCode() == KeyCode.V) {
+                handlePasteAction();  // Call paste handler
```
[the remark this answers is not in the document]
Comment (MrunalKashid02) [reply src/main/java/org/jabref/gui/fieldeditors/KeywordsEditor.java:98]: Yeah i made mistake previous by use ai. but the changes i am doing by understanding codebase and as per  suggestion i didnt take help from ai for code. I will make changes.

**Comment by MrunalKashid02** [comment]
I am getting this on mail what does it mean

Screenshot 2024-10-22 114558

**Comment by koppor** [comment]
> I am getting this on mail what does it mean

That you forked JabRef and "our" workflows are running at your side, too.

- I tried to disable them for forks (also for you then) - https://github.com/JabRef/jabref/pull/12097
- I field an upstream issue https://github.com/takanome-dev/assign-issue-action/issues/251

You need to update the main branch in your fork, then the error should be gone!

**Comment by subhramit** [comment]
@MrunalKashid02 hey, are you still working on it? You can otherwise close the PR.

File: src/main/java/org/jabref/gui/fieldeditors/KeywordsEditor.java [hunk 4]
```diff
@@ -86,16 +93,84 @@ public KeywordsEditor(Field field,
...
+        });
 
         Bindings.bindContentBidirectional(keywordTagsField.getTags(), viewModel.keywordListProperty());
     }
 
+    private void handlePasteAction() {
+            String pastedText = ClipBoardManager.getContents();
+
+            if (!pastedText.isEmpty()) {
+                String[] keywords = pastedText.split(String.valueOf(viewModel.getKeywordSeparator()));
```
[the remark this answers is not in the document]
Comment (MrunalKashid02) [reply src/main/java/org/jabref/gui/fieldeditors/KeywordsEditor.java:116]: Can elobarate more how this can be done

**Pull request closed without merging**